### ISSUE TITLE
Exploring implementation using Matplotlib as backend

### ISSUE THREAD

**Issue by kakirastern** [astropy/astrowidgets#73]
I am interested in exploring the development of a non-interactive version using Matplotlib as the backend, towards a non-interactive version which would implement as much of the API as possible in a non-interactive format. However, I have never done anything similar. Is there any suggestions as to some workflow or just simple guidelines I should follow?

**Comment by pllim** [comment]
Thank you for your interest!

I can't say if mine is the _suggested_ workflow, but personally, I would start with `example_notebook/ginga_widget.ipynb`. For prototyping, it would be easier to implement your version of `astrowidgets.core.ImageWidget` in your notebook cell. That way, you don't need to restart kernel or whatever to install your dev version constantly.

So, in your prototype, instead of `from astrowidgets import ImageWidget`, you would be using a local version of the `ImageWidget`. The goal is to replicate that notebook using your own implementation without changing the public API.

Public methods that are not working yet can raise `NotImplementedError` for now but they cannot be deleted.

However, optional keyword does not need to exist in your implementation but invoking them also cannot cause error. So a call like `ImageWidget(logger=logger)` -- while you do not have to necessarily do anything useful with the logger at this point, you have to support that API, so a `**kwargs` in your constructor would solve that problem right away.

@mwcraig or @eteq , feel free to correct me if I am steering @kakirastern in the wrong direction.

**Comment by kakirastern** [comment]
Thanks @pllim for the info!

**Comment by mwcraig** [comment]
This seems like a good way to go. It wouldn't be a bad approach to continue to use ginga to generate the underlying image but then display the image with matplotlib....though I suppose that the classes/functions in astropy.visualization helps quite a bit with converting a raw astronomical image to something displayable by matplotlib. 

There is a pretty limited implementation of a simple image viewer using matplotlib at https://github.com/mwcraig/ccd-reduction-and-photometry-guide/blob/master/notebooks/convenience_functions.py#L6

https://github.com/mwcraig/ccd-reduction-and-photometry-guide/blob/master/notebooks/convenience_functions.py#L6 motivation for wanting a non-interactive backend like matplotlib is that the API written for the image widget is really convenient for setting up static plots too!

**Comment by pllim** [comment]
At one point, outside of notebook, I wrote a script to grab mouse click X & Y locations out of matplotlib imshow. But AFAIK, there is no way to natively have matplotlib then convert X & Y to RA & DEC, so that is something you need to re-implement using `astropy.wcs`.

**Comment by kakirastern** [comment]
Thanks for the leads, I will definitely follow up on those

**Comment by kakirastern** [comment]
I really like the idea of using Ginga to render directly into a Matplotlib figure, even though support is limited to viewing widget itself. The strength would definitely be in the overplot options.

**Comment by kakirastern** [comment]
BTW, I might as well just call my version of the widget `MatImageWidget`...

**Comment by mwcraig** [comment]
Maybe `MattImageWIdget`? 😉

**Comment by pllim** [comment]
Matplotlib is usually shortened to MPL.

**Comment by pllim** [comment]
I don't know what is the benefit of using this widget if it requires Ginga anyway. Ginga also can overplot by overlaying its canvas objects. IMHO I would use this if I want publication quality plots without having to install Ginga. Matt, what is the use case that you have in your mind for this implementation?

**Comment by kakirastern** [comment]
Ah, I see... then I will go without Ginga then, just Matplotlib.

**Comment by michiboo** [comment]
I have try to just use Matplotlib but notice the quality of the image was not as good.

**Comment by kakirastern** [comment]
Hi @michiboo, would it be possible to share a bit of the details about your Matplotlib settings for the displaying of the images? Would like to see if I will get the same qualify with my setup or if I can reproduce the "qualify problem".

**Comment by kakirastern** [comment]
Matplotlib can produce quite good resolution images from my experience...

**Comment by michiboo** [comment]
If you compare with one produce by Ginga, you can see the difference.

**Comment by kakirastern** [comment]
Okay, thanks @michiboo. I will keep that in mind in case the issue becomes a problem.

**Comment by pllim** [comment]
You can set different Matplotlib backends, as with Ginga. By default, I think Ginga uses Qt, while Matplotlib uses TkAgg. If you want, you can play with different backends and see if they make any, well, difference.

https://matplotlib.org/faq/usage_faq.html#what-is-a-backend

**Comment by kakirastern** [comment]
Yup, sounds like some good experimentation to try out... I use TkAgg as the default Matplotlib backend. But I can certainly change it to something else and see the difference.

**Comment by kakirastern** [comment]
I think I will try working on this enhancement a bit since I would like to learn about widgets... But I have decided to use astroquery as my GSoC 2019 project idea, since that one is more flexible and I can adjust my workload based on my progress. Turned out a lot of work is needed to develop astroquery a bit more, but help is lacking. And, I see some other potential candidates have already expressed interest in the astrowidgets idea already, so I think it would be better for me to choose something else. That way maybe our collective chances of getting selected will be greater. That's just a guess though, based on my understanding of the situation.

**Comment by pllim** [comment]
@kakirastern , you should double check on the official GSoC mailing list for Open Astronomy. I think Open Astronomy only has a limited number of spots. Not all proposed projects will be assigned a participant. Please check over there before making a final decision.

**Comment by kakirastern** [comment]
Okay @pllim, I will give both ideas a try over the next week or so, and see how things go before making a final decision then.

**Comment by kakirastern** [comment]
Hi @pllim, I modified a bit of your matplotlib script to make it work in the tentative `MatImageWidget` as follows:
```
def plot_click(data):
    """Plot dummy data and handle user clicks."""

    # Show data
    if data is not False:
        plt.imshow(data)

    def _on_click(event):
        """Print and mark clicked coordinates."""

        # Print data coordinates to screen
        print('X, Y:', event.xdata, event.ydata)

        # Mark them on plot too
        plt.plot(event.xdata, event.ydata, 'rx')

    # Register click events
    plt.connect('button_press_event', _on_click)

    # Show plot
    plt.show()
```

And it is indeed working for on IPython now if I do the following:
```
import matplotlib  
matplotlib.use('TkAgg') 
  
from astrowidgets_matplotlib_core import MatImageWidget  
import matplotlib.pyplot as plt
w = MatImageWidget  
w.plot_click(([1, 2, 3, 4], [1, 4, 9, 16]))
```

Then I would get the image:
![Figure_1](https://user-images.githubusercontent.com/26264600/54244922-ba3db680-4569-11e9-8a03-d9d46d152205.png)

At least things are functional for now. Will need to tweak the code some more to make it good.

**Comment by pllim** [comment]
Do you mean `w = MatImageWidget()` (the instance, not the class)?

This is a good start. At some point, the workflow would be something like this:

```python
w = MatImageWidget()
w.load_array([[1, 2, 3, 4], [1, 4, 9, 16]])
w  # In a notebook, this would display the widget
```

Then on the widget, you can click and have the click do something (like display X and Y in a status bar).

Please note that I am not asking you do implement all this immediately, but rather give you the big picture, so you know what implementation is suitable and what is not. Keep up the good work!

**Comment by kakirastern** [comment]
Yup, sure. Now I see...

**Comment by kakirastern** [comment]
And, borrowing @mwcraig's script for show image, if I do the following:
```
from astrowidgets_matplotlib_core import MatImageWidget 
from astropy.io import fits

w = MatImageWidget()

filename = 'https://astropy.stsci.edu/data/photometry/spitzer_example_image.fits'  
numhdu = 0
w.load_fits(filename, 0, None)
hdul = fits.open(filename)
data = hdul[0].data
hdr = hdul[0].header
w.show_image(data)
```
I would get an image like this: 

![Figure_2](https://user-images.githubusercontent.com/26264600/54452367-f8b4ba80-478f-11e9-9691-bbf3247baf98.png)

I think I will need to tweak the size of the colorbar relative to the figsize a bit so that it will fit the figure size for each setting better, like the one I have shown above.

**Comment by kakirastern** [comment]
Definitely should work on making the widget "pan" and "zoon" very soon...

**Comment by kakirastern** [comment]
As an experiment: If I use "auto" aspect as an argument for imshow, I found that generally I would get the height of the figure the same as that of the colorbar, as follows... 
![Figure_2_auto_aspect](https://user-images.githubusercontent.com/26264600/54463708-e5fdae00-47ae-11e9-9cad-bef87ee76c6b.png)
Is this desirable behavior?

**Comment by kakirastern** [comment]
Or alternatively I can use my preferred `aspect="equal"` in `imshow()`, which gives something very similar to the original effect, like the following: 

![Figure_2_aspect_equal](https://user-images.githubusercontent.com/26264600/54464045-1bef6200-47b0-11e9-8cd7-44d4c1338939.png)

Anyway, for generating the above two plots I am digging the constraints set by the attributes `fraction=0.046, pad=0.04` in colorbar(). I think I prefer using an aspect setting in `imshow()` instead of in `colorbar()`, if they give the same effects... Any thoughts or preferences for me to use?

**Comment by kakirastern** [comment]
After using the `wcs` from the `header` of the same FITS file as `projection`, I was able to get the X and Y axis into RA and Dec in "dd:mm:ss" format so that the image now looks like the following instead (with `aspect='auto'`): 

![Figure_3_aspect_auto](https://user-images.githubusercontent.com/26264600/54465110-84404280-47b4-11e9-8082-63e1c350addb.png)

**Comment by kakirastern** [comment]
Can get the Matplotlib canvas working in an interactive mode in IPython... not sure if this would help but it looks cool 😮:

![Screenshot 2019-03-16 14 20 22](https://user-images.githubusercontent.com/26264600/54471613-c2f6ec80-47f6-11e9-8aa2-fc7b5f208642.png)

I understand I will need to incorporate Jupyter widgets somehow...
